File: scaffold/src/main/java/com/vgaw/scaffold/page/common/PermissionAc.java
package com.vgaw.scaffold.page.common;

import android.Manifest;
import android.app.Activity;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.graphics.Color;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.app.ActivityCompat;

import com.vgaw.scaffold.R;
import com.vgaw.scaffold.page.ScaffoldAc;
import com.vgaw.scaffold.page.ScaffoldDialog;
import com.vgaw.scaffold.page.ReqCodeConstant;
import com.vgaw.scaffold.util.dialog.DialogUtil;
import com.vgaw.scaffold.util.phone.SystemPageUtil;
import com.vgaw.scaffold.util.statusbar.StatusBarUtil;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class PermissionAc extends ScaffoldAc {
    private static Map<String, String> sPermissionDesMap = new HashMap<>();
    private Map<String, Boolean> mPermissionDeniedMap;

    private String[] mPermissionArray;
    private String mDes;
    private boolean mCheck;

    static {
        sPermissionDesMap.put(Manifest.permission.WRITE_EXTERNAL_STORAGE, "存储位置");
        sPermissionDesMap.put(Manifest.permission.RECORD_AUDIO, "麦克风");
        sPermissionDesMap.put(Manifest.permission.READ_PHONE_STATE, "设备信息");
        sPermissionDesMap.put(Manifest.permission.CAMERA, "相机");
    }

    public static boolean hasPermission(Activity activity, String[] permissionArray) {
        for (String item : permissionArray) {
            if (ActivityCompat.checkSelfPermission(activity, item) != PackageManager.PERMISSION_GRANTED) {
                return false;
            }
        }
        return true;
    }

    public static void startActivityForResult(Activity activity, int reqCode, String[] perssionArray) {
        startActivityForResult(activity, reqCode, perssionArray, null);
    }

    public static void startActivityForResult(Activity activity, int reqCode, String[] perssionArray, String des) {
        Intent intent = new Intent(activity, PermissionAc.class);
        intent.putExtra("permission_array", perssionArray);
        intent.putExtra("des", des);
        activity.startActivityForResult(intent, reqCode);
        activity.overridePendingTransition(0, 0);
    }

    @Override
    public void finish() {
        super.finish();
        overridePendingTransition(0, 0);
    }

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.permission_ac);
        StatusBarUtil.setColor(this, Color.TRANSPARENT);

        Intent intent = getIntent();
        mPermissionArray = intent.getStringArrayExtra("permission_array");
        mDes = intent.getStringExtra("des");
        requestPermission();
    }

    @Override
    protected void onResume() {
        super.onResume();
        if (mCheck) {
            mCheck = false;
            if (requestPermission()) {
                onSuc();
            }
        }
    }

    @Override
    protected void onStop() {
        super.onStop();
        mCheck = true;
    }

    @Override
    public void onRequestPermissionsResult(int requestCode, @NonNull String[] permissions, @NonNull int[] grantResults) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults);
        if (requestPermission()) {
            onSuc();
        }
    }

    @Override
    public void onBackPressed() {
        //super.onBackPressed();
        onFail();
    }

    private void onSuc() {
        Intent intent = getIntent();
        setResult(Activity.RESULT_OK, intent);
        finish();
    }

    private void onFail() {
        Intent intent = getIntent();
        setResult(Activity.RESULT_CANCELED, intent);
        finish();
    }

    private boolean requestPermission() {
        StringBuilder sb = new StringBuilder();
        List<String> permissionList = new ArrayList<>();
        boolean hasAlwaysDenied = false;
        for (String item : mPermissionArray) {
            if (proPermission(getSelf(), item, permissionList, sb)) {
                hasAlwaysDenied = true;
            }
        }
        if (permissionList.size() > 0) {
            String des = (mDes == null ? String.format("为了正常使用%s，请授予以下权限", getString(R.string.app_name)) : mDes);
            ScaffoldDialog baseDialog = ScaffoldDialog.newInstance(String.format("%s：\n%s", des,
                    sb.substring(0, sb.length() - 1)), "取消", hasAlwaysDenied ? "前往设置" : "开始授权");
            baseDialog.setOnLeftBtnClickListener(v -> {
                DialogUtil.dismissDialog(getSupportFragmentManager(), "permission_dialog");

                onFail();
            });
            boolean finalHasAlwaysDenied = hasAlwaysDenied;
            baseDialog.setOnRightBtnClickListener(v -> {
                DialogUtil.dismissDialog(getSupportFragmentManager(), "permission_dialog");

                if (finalHasAlwaysDenied) {
                    SystemPageUtil.jump2PermissoinPage(getSelf());
                } else {
                    String[] tempArray = new String[permissionList.size()];
                    permissionList.toArray(tempArray);
                    ActivityCompat.requestPermissions(getSelf(),
                            tempArray, ReqCodeConstant.PERMISSION);
                }
            });
            DialogUtil.showDialog(baseDialog, getSupportFragmentManager(), "permission_dialog");
            return false;
        }
        return true;
    }

    /**
     * 对于拒绝的权限，是否被永久拒绝（不再提示）
     * @param activity
     * @param permission
     * @param permissionList
     * @param sb
     * @return
     */
    private boolean proPermission(Activity activity, String permission, List<String> permissionList, StringBuilder sb) {
        if (ActivityCompat.checkSelfPermission(activity, permission) != PackageManager.PERMISSION_GRANTED) {
            sb.append(sPermissionDesMap.get(permission)).append("，");
            permissionList.add(permission);

            if (mPermissionDeniedMap == null) {
                mPermissionDeniedMap = new HashMap<>();
            }
            Boolean deniedBefore = mPermissionDeniedMap.get(permission);
            boolean deniedNow = ActivityCompat.shouldShowRequestPermissionRationale(activity, permission);
            if (deniedBefore == null) {
                mPermissionDeniedMap.put(permission, deniedNow);
            } else {
                if (!deniedBefore && !deniedNow) {
                    return true;
                }
            }
        }
        return false;
    }
}

File: scaffold/src/main/java/com/vgaw/scaffold/page/common/PermissionAc.kt
package com.vgaw.scaffold.page.common

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.content.pm.PackageManager
import android.graphics.Color
import android.os.Bundle
import android.view.View
import androidx.core.app.ActivityCompat
import com.vgaw.scaffold.R
import com.vgaw.scaffold.page.ScaffoldAc
import com.vgaw.scaffold.page.ScaffoldDialog
import com.vgaw.scaffold.util.dialog.DialogUtil
import com.vgaw.scaffold.util.phone.SystemPageUtil
import com.vgaw.scaffold.util.statusbar.StatusBarUtil
import timber.log.Timber

/**
 * todo
 * 1. 多个权限同一个描述；
 * 2. 不分手机功能异常；
 * 3. 先申请权限再弹出描述；
 *
 * 1. 读写权限自动适配
 */
class PermissionAc : ScaffoldAc() {
    companion object {
        private const val REQ_CODE_PERMISSION = 0

        private val sPermissionDesMap = HashMap<String, String>().apply {
            this[Manifest.permission.READ_EXTERNAL_STORAGE] = "读取文件"
            this[Manifest.permission.WRITE_EXTERNAL_STORAGE] = "存储文件"
            this[Manifest.permission.RECORD_AUDIO] = "麦克风"
            this[Manifest.permission.READ_PHONE_STATE] = "设备信息"
            this[Manifest.permission.CAMERA] = "相机"
            this[Manifest.permission.ACCESS_COARSE_LOCATION] = "粗略位置信息"
            this[Manifest.permission.ACCESS_FINE_LOCATION] = "精确位置信息"
            this[Manifest.permission.CALL_PHONE] = "拨打电话"
            this[Manifest.permission.USE_SIP] = "使用SIP"
            this[Manifest.permission.WRITE_CONTACTS] = "修改通讯录"
            this[Manifest.permission.ACCESS_BACKGROUND_LOCATION] = "后台获取位置【请点击始终允许】"
        }

        fun startActivityForResult(activity: Activity, reqCode: Int, perssionArray: Array<String>): Boolean {
            if (!hasPermission(activity, perssionArray)) {
                startActivityForResult(activity, reqCode, perssionArray, null)
                return false
            }
            return true
        }

        fun startActivityForResult(activity: Activity, reqCode: Int, perssionArray: Array<String>, des: String?): Boolean {
            if (!hasPermission(activity, perssionArray)) {
                val intent = Intent(activity, PermissionAc::class.java)
                intent.putExtra("permission_array", perssionArray)
                if (des != null) {
                    intent.putExtra("des", des)
                }
                activity.startActivityForResult(intent, reqCode)
                activity.overridePendingTransition(0, 0)
                return false
            }
            return true
        }

        fun hasPermission(activity: Activity, permissionArray: Array<String>): Boolean {
            for (item in permissionArray) {
                if (ActivityCompat.checkSelfPermission(activity, item) != PackageManager.PERMISSION_GRANTED) {
                    return false
                }
            }
            return true
        }
    }

    private var mPermissionDeniedMap: HashMap<String, Boolean>? = null

    private lateinit var mPermissionArray: Array<String>
    private var mDes: String? = null

    private var mCheck: Boolean = false

    override fun finish() {
        super.finish()
        overridePendingTransition(0, 0)
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.permission_ac)
        StatusBarUtil.setColor(this, Color.TRANSPARENT)

        val intent = intent
        mPermissionArray = intent.getStringArrayExtra("permission_array")!!
        mDes = intent.getStringExtra("des")
        requestPermission()
    }

    override fun onResume() {
        super.onResume()
        if (mCheck) {
            mCheck = false
            if (requestPermission()) {
                onSuc()
            }
        }
    }

    override fun onStop() {
        super.onStop()
        mCheck = true
    }

    override fun onRequestPermissionsResult(requestCode: Int, permissions: Array<out String>, grantResults: IntArray) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)
        if (requestPermission()) {
            onSuc()
        }
    }

    override fun onBackPressed() {
        //super.onBackPressed()
        onFail()
    }

    private fun onSuc() {
        val intent = intent
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    private fun onFail() {
        val intent = intent
        setResult(Activity.RESULT_CANCELED, intent)
        finish()
    }

    private fun requestPermission(): Boolean {
        val permissionList = ArrayList<String>()
        var hasAlwaysDenied = false
        for (item in mPermissionArray) {
            if (proPermission(getSelf(), item, permissionList)) {
                hasAlwaysDenied = true
            }
        }
        if (permissionList.size > 0) {
            val sb = StringBuilder()
            permissionList.forEach { sb.append(sPermissionDesMap[it]).append("，") }

            val des = mDes ?: String.format("为了正常使用%s，请授予以下权限", getString(R.string.app_name))
            val baseDialog = ScaffoldDialog.newInstance(String.format("%s：\n%s", des, sb.substring(0, sb.length - 1)), "取消", /*if hasAlwaysDenied "前往设置" else */"开始授权")

            baseDialog.setOnLeftBtnClickListener(View.OnClickListener {
                DialogUtil.dismissDialog(supportFragmentManager, "permission_dialog")

                onFail()
            })
            val finalHasAlwaysDenied = hasAlwaysDenied
            baseDialog.setOnRightBtnClickListener(View.OnClickListener {
                DialogUtil.dismissDialog(supportFragmentManager, "permission_dialog")

                if (finalHasAlwaysDenied) {
                    SystemPageUtil.jump2PermissoinPage(getSelf())
                } else {
                    val tempList = mutableListOf<String>()
                    for (i in 0 until permissionList.size) {
                        tempList.add(permissionList[i])
                        tempList.add(sPermissionDesMap.get(permissionList[i])!!)
                    }
                    ActivityCompat.requestPermissions(getSelf(),
                        tempList.toTypedArray(), REQ_CODE_PERMISSION)
                }
            })
            DialogUtil.showDialog(baseDialog, supportFragmentManager, "permission_dialog")
            return false
        }
        return true
    }

    /**
     * 对于拒绝的权限，是否被永久拒绝（不再提示）
     * @param activity
     * @param permission
     * @param permissionList
     * @param sb
     * @return
     */
    private fun proPermission(activity: Activity, permission: String, permissionList: ArrayList<String>): Boolean {
        val check = ActivityCompat.checkSelfPermission(activity, permission)
        if (Manifest.permission.WRITE_EXTERNAL_STORAGE.equals(permission)) {
            Timber.d("grant: %s", check)
        }
        if (check != PackageManager.PERMISSION_GRANTED) {
            permissionList.add(permission)

            if (mPermissionDeniedMap == null) {
                mPermissionDeniedMap = HashMap()
            }
            val deniedBefore = mPermissionDeniedMap!!.get(permission)
            val deniedNow = ActivityCompat.shouldShowRequestPermissionRationale(activity, permission)
            if (deniedBefore == null) {
                mPermissionDeniedMap!!.put(permission, deniedNow)
            } else {
                if (!deniedBefore && !deniedNow) {
                    return true
                }
            }
        }
        return false
    }
}
